core/src/main/java/com/sequenceiq/cloudbreak/service/publicendpoint/FreeIPAEndpointManagementService.java

```diff
@@ -0,0 +1,128 @@
+
+import com.sequenceiq.cloudbreak.auth.ThreadBasedUserCrnProvider;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.loadbalancer.LoadBalancer;
+import com.sequenceiq.cloudbreak.service.LoadBalancerConfigService;
+import com.sequenceiq.cloudbreak.service.environment.EnvironmentClientService;
+import com.sequenceiq.cloudbreak.service.stack.LoadBalancerPersistenceService;
+import com.sequenceiq.common.api.type.LoadBalancerType;
+import com.sequenceiq.freeipa.api.v1.dns.DnsV1Endpoint;
+import com.sequenceiq.freeipa.api.v1.dns.model.AddDnsARecordRequest;
+import com.sequenceiq.freeipa.api.v1.dns.model.AddDnsCnameRecordRequest;
+
+@Service
+public class FreeIPAEndpointManagementService extends BasePublicEndpointManagementService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIPAEndpointManagementService.class);
+
+    private static final String DOMAIN_PART_DELIMITER = ".";
+
+    @Inject
+    private DnsV1Endpoint dnsV1Endpoint;
+
+    @Inject
+    private EnvironmentClientService environmentClientService;
+
+    @Inject
+    private LoadBalancerPersistenceService loadBalancerPersistenceService;
+
+    @Inject
+    private LoadBalancerConfigService loadBalancerConfigService;
+
+    public void registerLoadBalancerDomainWithFreeIPA(Stack stack) {
+        Optional<LoadBalancer> loadBalancerOptional = getLoadBalancer(stack.getId());
+        if (loadBalancerOptional.isPresent()) {
+            LoadBalancer loadBalancer = loadBalancerOptional.get();
+            try {
+                if (StringUtils.isNotEmpty(loadBalancer.getDns())) {
+                    sendAddDnsCnameRecordRequest(stack, loadBalancer);
+                } else if (StringUtils.isNotEmpty(loadBalancer.getIp())) {
+                    sendAddDnsARecordRequest(stack, loadBalancer);
+                } else {
+                    LOGGER.error("Unable to find DNS or IP for load balancer. Load balancer will not be registered with FreeIPA.");
+                }
+            } catch (Exception e) {
+                LOGGER.error("Unable to register load balancer with FreeIPA", e);
+            }
+        } else {
+            LOGGER.debug("No load balancer found that needs to be registered with FreeIPA");
+        }
+    }
+
+    private void sendAddDnsCnameRecordRequest(Stack stack, LoadBalancer loadBalancer) {
+        String targetFQDN = StringUtils.appendIfMissing(loadBalancer.getDns(), DOMAIN_PART_DELIMITER);
+        String endpoint = loadBalancer.getEndpoint();
+
+        AddDnsCnameRecordRequest request = new AddDnsCnameRecordRequest();
+        request.setCname(endpoint);
+        request.setTargetFqdn(targetFQDN);
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        LOGGER.debug("Registering load balancer with target FQDN {} in FreeIPA with CNAME {}", targetFQDN, endpoint);
+        ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsCnameRecord(request));
```

Comment:
this should call `addDnsCnameRecordInternal`

Reply:
```
String accountId = ThreadBasedUserCrnProvider.getAccountId();
ThreadBasedUserCrnProvider.doAsInternalActor(() -> dnsV1Endpoint.addDnsCnameRecordInternal(accountId, request));
```